src/scala/org/pantsbuild/scoverage/report/ScoverageReport.scala

```diff
@@ -0,0 +1,254 @@
+    targetFilters: Seq[String]) {
+    val writeReports: Boolean = {
+      val write = writeHtml || writeXml
+      if (!write) {
+        // we could end here, but we will still attempt to load the coverage and dump stats as info
+        // to the log, so maybe somebody would want to run this w/o generating reports?
+        logger.warn("No report output format specified, so no reports will be written.")
+      }
+      write
+    }
+  }
+
+  object ReportOptions {
+    def apply(settings: Settings): ReportOptions = {
+      ReportOptions(
+        loadDataDir = settings.loadDataDir,
+        measurementsPath = settings.measurementsDirPath,
+        sourcePath = settings.sourceDirPath,
+        reportPath = settings.reportDirPath,
+        dataPath = settings.dataDirPath,
+        writeHtml = settings.writeHtmlReport,
+        writeXml = settings.writeXmlReport,
+        writeXmlDebug = settings.writeXmlDebug,
+        cleanOld = settings.cleanOldReports,
+        targetFilters = settings.targetFilters)
+    }
+  }
+  /**
+   *
+   * @param dataDirs list of measurement directories for which coverage
+   *                 report has to be generated
+   * @return         [Coverage] object for the [dataDirs]
+   */
+  private def aggregatedCoverage(dataDirs: Seq[File]): Coverage = {
+    var id = 0
+    val coverage = Coverage()
+    dataDirs foreach { dataDir =>
+      val coverageFile: File = Serializer.coverageFile(dataDir)
+      if (coverageFile.exists) {
+        val subcoverage: Coverage = Serializer.deserialize(coverageFile)
+        val measurementFiles: Array[File] = IOUtils.findMeasurementFiles(dataDir)
+        val measurements = IOUtils.invoked(measurementFiles.toIndexedSeq)
+        subcoverage.apply(measurements)
+        subcoverage.statements foreach { stmt =>
+          // need to ensure all the ids are unique otherwise the coverage object will have stmt collisions
+          id = id + 1
+          coverage add stmt.copy(id = id)
+        }
+      }
+    }
+    coverage
+  }
+
+  /**
+   *
+   * @param dataDirs list of measurement directories for which coverage
+   *                 report has to be generated
+   * @return         Coverage object wrapped in Option
+   */
+  private def aggregate(dataDirs: Seq[File]): Option[Coverage] = {
+    logger.info(s"Found ${dataDirs.size} subproject scoverage data directories [${dataDirs.mkString(",")}]")
+    if (dataDirs.nonEmpty) {
+      Some(aggregatedCoverage(dataDirs))
+    } else {
+      None
+    }
+  }
+
+  /**
+   *
+   * @param dataDir root directory to search under
+   * @return        all the directories and subdirs containing scoverage files beginning at [dataDir]
+   */
+  def getAllCoverageDirs(dataDir: File, acc: Seq[File]): Seq[File] = {
+    if (dataDir.listFiles.filter(_.isFile).toSeq.exists(_.getName contains "scoverage.coverage")) {
+      dataDir.listFiles.filter(_.isDirectory).toSeq
+        .foldRight(acc :+ dataDir) { (e, a) => getAllCoverageDirs(e, a) }
```

Comment:
Appending to a `Seq` is not always cheap. If you want to ensure that you have cheap append, your argument should probably be a `Vector[File]`.

Reply:
I assume prepending is cheap?

Reply:
Prepending is only cheap if you are using `List` or `Vector`... but not `Array`, for example. So it's good to be specific.

Reply:
Alrighty, made it to a vector. Thanks